### PR TITLE
Add the tempDir as parameter

### PR DESCRIPTION
**Pull request by TopperDEL**
What: 
When creating a new Uplink I wanted to provide a temp-directory to use - this is necessary at least on Android.

Why:
Otherwise Uploads (and presumably Downloads) on Android won't work.

Please describe the tests:
Tested it with my .Net-Wrapper (uplink.NET).
 
Please describe the performance impact:
Should be None.

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

**Comment by mniewrzal** [comment]
@TopperDEL thanks for contribution, to make this change complete you need also update automatic tests https://build.dev.storj.io/blue/organizations/jenkins/storj/detail/PR-3065/1/tests

**Comment by TopperDEL** [comment]
I've corrected the automated test, but it failed again. To be honest, I don't know what went wrong. My Change should not Impact what I see in the test-fail-log.

Can someone have a look at it and tell me if I Need to adjust my PR any further?

**Comment by mniewrzal** [comment]
@bryanchriswhite can you take a look and give some help to @TopperDEL ?

**Comment by bryanchriswhite** [comment]
:eyes:

**Comment by bryanchriswhite** [comment]
@TopperDEL try merging https://github.com/TopperDEL/storj/pull/2 and see if that fixes jenkins.

**Comment by bryanchriswhite** [comment]
I'm still looking into this. I'm not sure yet of the root cause of the backwards-compatibility check failure.

**Comment by bryanchriswhite** [comment]
I am unable to reproduce this error locally. When I run:
```
psql -U postgres -c 'create database teststorj3;' && STORJ_SIM_POSTGRES=postgres://postgres@localhost/teststorj3?sslmode=disable make test-sim-backwards-compatible
```
I don't see the `Bucket not empty: bucket-123` error that shows up at the end of jenkins' backwards-compatibility check log.

**Comment by TopperDEL** [comment]
I'm sorry that I cannot help you out on this.

**Comment by mniewrzal** [comment]
@TopperDEL @bryanchriswhite it looks that failing test was some kind of instability around master. Now its passing.

**Comment by bryanchriswhite** [comment]
> @TopperDEL @bryanchriswhite it looks that failing test was some kind of instability around master. Now its passing.

Hooray!

File: lib/uplinkc/uplink.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ type Uplink struct {
 // an error in cerr, when there is one.
 //
 // Caller must call close_uplink to close associated resources.
-func new_uplink(cfg C.UplinkConfig, cerr **C.char) C.UplinkRef {
-	scope := rootScope("") // TODO: pass in as argument
+func new_uplink(cfg C.UplinkConfig, tempDir *C.char, cerr **C.char) C.UplinkRef {
```
Comment (mniewrzal) [inline lib/uplinkc/uplink.go]: I'm just wondering if we should make `tempDir` part of `UplinkConfig` to avoid keeping it as a separate parameter.

Comment (TopperDEL) [reply lib/uplinkc/uplink.go]: Sounds reasonable

Comment (bryanchriswhite) [reply lib/uplinkc/uplink.go]: My hesitation with that is that besides breaking symmetry with the go struct, the uplink config  is still not well defined as it currently contains a single `Volatile` struct. Additionally, keeping it as a parameter makes it more visible and clear that it's required (which apparently it is).

**Pull request merged**